sdks/python/apache_beam/runners/worker/bundle_processor.py

```diff
@@ -1005,18 +995,97 @@ def process_bundle(self, instruction_id):
         assert timer_info.output_stream is not None
         timer_info.output_stream.close()
 
-      return ([
-          self.delayed_bundle_application(op, residual) for op,
-          residual in execution_context.delayed_applications
-      ],
-              self.requires_finalization())
+      if requires_finalization is None:
+        requires_finalization = self.requires_finalization()
+
+      if delayed_applications is None:
+        delayed_applications = [self.delayed_bundle_application(op, residual)
+                                for op, residual in
+                                execution_context.delayed_applications]
+
+      return delayed_applications, requires_finalization
 
     finally:
       # Ensure any in-flight split attempts complete.
       with self.splitting_lock:
         pass
       self.state_sampler.stop_if_still_running()
 
+  def maybe_process_remotely(self,
+                             data_channels,  # type: DefaultDict[DataChannel, list[str]]
+                             instruction_id,  # type: str
+                             input_op_by_transform_id,  # type: Dict[str, DataInputOperation]
+                             use_task_worker=True  # type: bool
+                             ):
+    # type: (...) -> Union[Tuple[None, None], Tuple[List[beam_fn_api_pb2.DelayedBundleApplication], bool]]
```

Comment:
@violalyu like-wise, how many times in the lifespan of a `BundleProcessor` would `maybe_process_remotely` be called?

Reply:
it would be called the same number of times as the number of bundles this current `BundleProcessor` is asked to process